### DISCUSSION TITLE
Create a database driver for edge environments like `@planetscale/database` or `@neondatabase/serverless`

### DISCUSSION THREAD

**Discussion by remorses** [supabase/supabase#15925] [Feature Requests] [answered]
I want to use my existing ORM instead of the `supabase` js client (postgrest) in an edge environment like Vercel edge functions, to do this i need a database driver that works with `fetch` instead of TCP connections 

Planetscale and Neon already have this feature:
https://github.com/neondatabase/serverless
https://github.com/planetscale/database-js

**Comment by kincaidoneil** [comment]
~~Similarly, I'd like to access my Supabase instance using Prisma within Vercel Edge functions, where TCP isn't supported.
It'd be great even for supabase-js to support raw SQL with PostgREST, so then we could easily implement our own HTTP-based driver adapter for Prisma (Prisma supports this extensibility now, but doesn't have one for Supabase).~~

See https://github.com/orgs/supabase/discussions/15925#discussioncomment-8573567

**Comment by kincaidoneil** [comment, marked as answer]
PSA: Supavisor [runs a WS proxy](https://github.com/supabase/supavisor/pull/82), **so Supabase is already compatible with edge environments like Vercel using [`neondatabase/serverless`](https://github.com/neondatabase/serverless)!**

Use Neon's package anywhere `node-postgres` is supported, and connect to Supabase with a Supavisor pooled string. Just set `neonConfig.pipelineConnect = false` 🎉

**Reply by ororsatti** [reply]
how?
how can you run directly from an edge function to prisma?

**Reply by kincaidoneil** [reply]
Example here: https://gist.github.com/kincaidoneil/bc2516111f0ec8850cd6020b8191b27b

**Reply by ororsatti** [reply]
`import { PrismaClient } from '@prisma/client'`

and this doesnt throw you an error by Deno?

theres a specific setup for using prisma in deno, according to the docs: https://docs.deno.com/runtime/tutorials/how_to_with_npm/prisma

**Reply by kincaidoneil** [reply]
This was primarily for Vercel Edge, not Deno. AFAIK Deno supports TCP so you don't need any of these edge driver workarounds, just instantiate Prisma normally?

**Reply by ororsatti** [reply]
that doesnt "jost" work sadly...

**Reply by mfchiz** [reply]
this is great. i was able to get the gist working. if i use the neon serverless driver are the web socket connections secured to the Supavisor proxy?

**Reply by kincaidoneil** [reply]
@mfchiz AFAIK the security model should be similar to connecting to a connection pooler like Supavisor over TCP. See Neon's config options [here](https://github.com/neondatabase/serverless/blob/main/CONFIG.md), by default it'll use TLS on the Postgres protocol layer.

**Reply by batuhanbilginn** [reply]
Thank you for the answer. It works on production but do you know how we can make this work in local supabase environment?

**Reply by kincaidoneil** [reply]
@batuhanbilginn Supabase CLI still uses the older connection pooler, PgBouncer, instead of Supavisor I think. Hopefully they update this. But, you can use [this script](https://gist.github.com/kincaidoneil/bc2516111f0ec8850cd6020b8191b27b#file-setup-local-env-ts) to start Postgres and Supavisor locally in Docker.

**Reply by batuhanbilginn** [reply]
Thank you so much for your script. I can confirm that Supabase CLI is now using Supavisor in the latest version. Even if I couldn't connect to the database on edge runtime so far just wanted to share it with you.

Also, there is an official guide for the cloud version of Supabase here: https://supabase.com/docs/guides/database/connecting-to-postgres/serverless-drivers

And, this is the issue I've created in CLI repo if you want to follow: https://github.com/supabase/cli/issues/2474

**Reply by Immortalin** [reply]
@batuhanbilginn can you document the compatibility with neon's drivers? Currently it's the only prisma compatible driver. None of the others work with prisma in an edge environment

**Reply by batuhanbilginn** [reply]
@Immortalin compatibility of what? I couldn't understand the question.

**Reply by Immortalin** [reply]
Compatibility of the neon driver with prisma and supabase.
